File: tuna-compiler/src/main/backend/abstract/IR.ts
import {Op, Schema, Utils } from '../ops/index';

export type Node<DATA={}, META extends "root" | "not root"="not root"> = DATA & {_meta: META}
export type ValueNode = PickNode<
    "Bool" | 
    "Object" | 
    "Comparison" | 
    "BoolAlg" | 
    "Int" | 
    "Saved" | 
    "String" |
    "FieldExists" |
    "Selection" |
    "GlobalObject" | 
    "Math" |
    "None" |
    "ArrayLiteral" |
    "Call" |
    "Keys" |
    "RoleInstance" |
    "GetType" |
    "Not" |
    "Is"
    >

export type Key = PickNode<"String" | "Saved">
export type AbstractNodes = PickNode<"GlobalObject">

export type BaseNodeDefs = {
    Return: Node<{value?: ValueNode}, "root">
    Bool: Node<{value: boolean}>
    DeleteField: Node<{}>,
    Field: Node<{key: Key, value: ValueNode}>
    Object: Node<{fields: PickNode<"Field">[]}>
    Int: Node<{value: number}> 
    None: Node<{}>,
    GetType: Node<{value: ValueNode}>
    Not: Node<{value: ValueNode}>
    Comparison: Node<
        {
        sign: "==" | "!=" | "<" | ">" | "<=" | ">="
        left: ValueNode
        right: ValueNode
    }>,
    Math: Node<{
        sign: "+" | "-" | "*" | "/",
        left: ValueNode
        right: ValueNode
    }>,
    Is: Node<{value: ValueNode, type: string}>
    BoolAlg: Node<{
        sign: "and" | "or", 
        left: ValueNode, 
        right: ValueNode}>

    Else: Node<{do: RootNode[]}>
    Finally: Node<{do: RootNode[]}>
    Conditional: Node<{cond: ValueNode, do: RootNode[]}>
    If: Node<{
        conditionally: PickNode<"Conditional" | "Finally" | "Else">[]
    }, "root">
    RoleInstance: Node<{
        role: Extract<Schema, {kind: "Role"}>,
        state?: PickNode<"Object">
    }>
    Noop: Node<{}, "root">
    Saved: Node<{arg: string}> 
    String: Node<{value: string}>,
    Selection: Node<{root: ValueNode, level: (ValueNode)[]}>
    FieldExists: Node<{value: ValueNode, field: Key}>
    Save: Node<{value: ValueNode, name: string}, "root">
    Update: Node<
    {
        root: PickNode<"Saved" | "GlobalObject">
        level: PickNode<"Selection">["level"]
        operation: PickNode<"DeleteField" | "Push"> | ValueNode,
    }, "root">
    GlobalObject: Node<{name: string}>
    ArrayForEach: Node<{target: ValueNode, do: RootNode[], arg: string}, "root">
    Keys: Node<{from: ValueNode}>
    ArrayLiteral: Node<{values: ValueNode[]}>
    Push: Node<{values: ValueNode[]}>
    Call: Node<{function_name: string, args: ValueNode[]}, "root">
}

export type NodeSet= {[K in string]: Node<{}, "not root" | "root">} 
type NodeInstance<S extends NodeSet, K extends keyof S> = Omit<S[K], "_meta"> & {kind: K}
export type AnyNodeFromSet<S extends NodeSet> = {
    [K in keyof S]: NodeInstance<S, K>
}[keyof S]

export type AnyNode = AnyNodeFromSet<BaseNodeDefs>
export type RootNode = AnyRootNodeFromSet<BaseNodeDefs>
export type AnyRootNodeFromSet<S extends NodeSet> = {
    [K in keyof S]: S[K]["_meta"] extends "not root" ? never : {kind: K} & Omit<S[K], "_meta">
}[keyof S]

export type PickNode<K extends keyof BaseNodeDefs> = Extract<AnyNode, {kind: K}>

File: tuna-compiler/rust/compiler/src/ir.rs

#![allow(unused_imports)]

use tuna_interpreter::data::{InterpreterType};
use tuna_interpreter::schemas::{Schema};

pub struct Conditional {
    pub condition: Value, 
    pub body: Vec<Root>
}

pub enum Mut {
    Overwrite(Value),
    Push(Vec<Value>),
    Delete
}

pub struct Call {
    pub function: String, 
    pub args: Vec<Value>
}

pub enum Root {
    Branch(Vec<Conditional>),
    Save {val: Value, name: String},
    Update {root: Saved, level: Vec<Value>, operation: Mut},
    ForEach {target: Value, body: Vec<Root>, arg: String},
    Call(Call),
    Return(Option<Value>)
}

pub struct Field {
    pub key: String, 
    pub value: Value
}
pub struct Saved(pub String);
type Value = Box<AnyValue>;
pub enum AnyValue {
    Bool(bool),
    Object(Vec<Field>),
    Int(i64),
    Double(f64),
    String(String),
    None,
    GetType(Value),
    Not(Value),
    BinaryOp {sign: Sign, left: Value, right: Value},
    Is {val: Value, typ: String},
    RoleInstance {schema: Schema, data: Vec<Field>},
    Saved(String),
    Selection {root: Value, level: Vec<Value>},
    Keys(Value),
    Array(Vec<Value>),
    Call(Call)
}

pub enum Sign {
    Eq,
    Neq,
    L,
    G,
    Leq,
    Geq,
    Plus,
    Minus,
    And,
    Or,
    Div,
    Mult
}


pub struct Function<'a> {
    pub name: &'a str,
    pub args: Vec<(Schema, String)>,
    pub body: Vec<Root>
}